test/test_datapipe.py

```diff
@@ -2268,5 +2274,78 @@ def test_old_dataloader(self):
         self.assertEqual(sorted(expected), sorted(items))
 
 
+class TestIterDataPipeSingletonConstraint(TestCase):
+
+    r"""
+    Each `IterDataPipe` can only have one active iterator. Whenever a new iterator is created, older
+    iterators are invalidated. These tests aim to ensure `IterDataPipe` follows this behavior.
+    """
+
+    def test_iterdatapipe_singleton_constraint(self):
+        r"""
+        Testing for the case where IterDataPipe's `__iter__` is a generator function.
+        """
+        # Functional Test: There should be an error with multiple iterators
+        source_dp: IterDataPipe = dp.iter.IterableWrapper(range(10))
+        it1 = iter(source_dp)
+        self.assertEqual(list(range(10)), list(it1))  # It can
+        it1 = iter(source_dp)
+        self.assertEqual(list(range(10)), list(it1))
+        it1 = iter(source_dp)
+        self.assertEqual(0, next(it1))
+        it2 = iter(source_dp)
+        self.assertEqual(0, next(it2))
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
+            next(it1)
+
+        r"""
+        Testing for the case where IterDataPipe's `__iter__` returns `self` and there is a `__next__` method
+        Note that the following DataPipe by is singleton by default (because `__iter__` returns `self`).
+        Therefore the test will merely make sure the calls return the same object.
+        """
+        class _CustomIterDP_Self(IterDataPipe):
+            def __init__(self, iterable):
+                self.iterable = iter(iterable)
+
+            def __iter__(self):
+                return self
+
+            def __next__(self):
+                return next(self.iterable)
+
+        # Functional Test: Check that every `__iter__` call returns the same object
+        source_dp = _CustomIterDP_Self(range(10))
+        it1 = iter(source_dp)
+        self.assertEqual(0, next(it1))
+        self.assertEqual(1, next(source_dp))
+        it2 = iter(source_dp)
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
+            next(it1)
+        with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
```

Comment:
wait, why it is invalid?

Reply:
do we have at least one valid iterator after this?

Reply:
In my original proposal, the users can are allowed to call `__iter__` as many times as they want. But after our offline discussion, we decided to apply the constraint on `__iter__` even when it may be returning `self`. This is because, having multiple active iterators referencing the same underlying object can lead to confusion and bugs. Plus the use case isn't really necessary when you can just reference `dp`. The code snippet below illustrates what the behavior would've been like if it is allowed:
```python
dp = CustomDP(range(10))
next(dp)  # returns 0
it1 = iter(dp)  # returns `self`
next(it1)  # returns 1
it2 = iter(dp)  # returns `self`
next(it2)  # returns 2
next(it1)  # returns 3
next(dp)  # returns 4
```

In the current implementation, calling `next(dp)` raises an error if there is more than 1 iterator created from that DataPipe. This is because the ID that tracks whether the iterator is valid (i.e. `dp._valid_iterator_id`) is tied to the DataPipe (and because it returns `self`, the ID is tied to the iterator as well). I think it has to be the case, otherwise when `next(dp)` is called. It has no idea to what ID to check against.

```python
source_dp = _CustomIterDP_Self(range(10))
it1 = iter(source_dp)
self.assertEqual(0, next(it1))
self.assertEqual(1, next(source_dp))
it2 = iter(source_dp)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(it1)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(it2)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(source_dp)
```

The simplest workaround for users is to re-define `dp` via `dp = CustomDP()` or avoid creating more than one iterator from `dp` by keep using `dp`  as the variable rather than calling `iter(dp)`.

Reply:
Here are the proposal you have
- For `DataPipe` returning `self` in `__iter__`: We disallow users to call `iter` multiple times.
- For `DataPipe` using a generator function as `__iter__`: We are making sure a new iterator is created and previous iterator is invalidated whenever `iter` is called.
As we are making iterator of `DataPiep` singleton (the same behavior), we are introducing a new difference on iterator. I think this is not preferable.

I think as we have `reset` function to help `DataPipe` to reset iterator. Why don't we rely on `reset` to clean up all buffer, etc. for `DataPipe` returning `self` in `__iter__`. And, the `IteratorDecorator` and `datapipe` itself should be able to track the iterator_id to invalidate previous iterator.

Reply:
I can see how that can be confusing.

I think this proposal will be best, given a DataPipe `dp` with `__next__` with iterators `it1` and `it2`:
1. Allows unlimited creation of iterators, but only the latest one is valid/active
2. Doesn't place restriction on `next(dp)` (but check for invalidation when `next(it1)` is called
    - We can't really place restriction on `next(dp)`, because any call to `next(it2)` gets delegated to `next(dp)` 
       within the `IteratorDecorator`

```python
source_dp = _CustomIterDP_Self(range(10))
it1 = iter(source_dp)
self.assertEqual(0, next(it1))
self.assertEqual(1, next(source_dp))
# Only invalidates `it1`, and not `source_dp`. Since methods of `it2` depends on `source_dp` remaining valid
it2 = iter(source_dp)
with self.assertRaisesRegex(RuntimeError, "This iterator has been invalidated"):
    next(it1)
self.assertEqual(0, next(it2))
self.assertEqual(1, next(source_dp))
```


Please examine this test for the full details:
https://github.com/pytorch/pytorch/blob/c77df3ecf5110560e602ff5c4b3b5825ffb168c0/test/test_datapipe.py#L2337-L2390